File: pulsar-tools/pulsar-browser/src/main/java/com/github/kklisura/cdt/protocol/v2023/commands/Animation.java
package com.github.kklisura.cdt.protocol.v2023.commands;

/*-
 * #%L
 * cdt-java-client
 * %%
 * Copyright (C) 2018 - 2023 Kenan Klisura
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */

import com.github.kklisura.cdt.protocol.v2023.events.animation.AnimationCanceled;
import com.github.kklisura.cdt.protocol.v2023.events.animation.AnimationCreated;
import com.github.kklisura.cdt.protocol.v2023.events.animation.AnimationStarted;
import com.github.kklisura.cdt.protocol.v2023.support.annotations.EventName;
import com.github.kklisura.cdt.protocol.v2023.support.annotations.Experimental;
import com.github.kklisura.cdt.protocol.v2023.support.annotations.ParamName;
import com.github.kklisura.cdt.protocol.v2023.support.annotations.Returns;
import com.github.kklisura.cdt.protocol.v2023.support.types.EventHandler;
import com.github.kklisura.cdt.protocol.v2023.support.types.EventListener;
import com.github.kklisura.cdt.protocol.v2023.types.runtime.RemoteObject;

import java.util.List;

@Experimental
public interface Animation {

  /** Disables animation domain notifications. */
  void disable();

  /** Enables animation domain notifications. */
  void enable();

  /**
   * Returns the current time of the an animation.
   *
   * @param id Id of animation.
   */
  @Returns("currentTime")
  Double getCurrentTime(@ParamName("id") String id);

  /** Gets the playback rate of the document timeline. */
  @Returns("playbackRate")
  Double getPlaybackRate();

  /**
   * Releases a set of animations to no longer be manipulated.
   *
   * @param animations List of animation ids to seek.
   */
  void releaseAnimations(@ParamName("animations") List<String> animations);

  /**
   * Gets the remote object of the Animation.
   *
   * @param animationId Animation id.
   */
  @Returns("remoteObject")
  RemoteObject resolveAnimation(@ParamName("animationId") String animationId);

  /**
   * Seek a set of animations to a particular time within each animation.
   *
   * @param animations List of animation ids to seek.
   * @param currentTime Set the current time of each animation.
   */
  void seekAnimations(
      @ParamName("animations") List<String> animations,
      @ParamName("currentTime") Double currentTime);

  /**
   * Sets the paused state of a set of animations.
   *
   * @param animations Animations to set the pause state of.
   * @param paused Paused state to set to.
   */
  void setPaused(
      @ParamName("animations") List<String> animations, @ParamName("paused") Boolean paused);

  /**
   * Sets the playback rate of the document timeline.
   *
   * @param playbackRate Playback rate for animations on page
   */
  void setPlaybackRate(@ParamName("playbackRate") Double playbackRate);

  /**
   * Sets the timing of an animation node.
   *
   * @param animationId Animation id.
   * @param duration Duration of the animation.
   * @param delay Delay of the animation.
   */
  void setTiming(
      @ParamName("animationId") String animationId,
      @ParamName("duration") Double duration,
      @ParamName("delay") Double delay);

  /** Event for when an animation has been cancelled. */
  @EventName("animationCanceled")
  EventListener onAnimationCanceled(EventHandler<AnimationCanceled> eventListener);

  /** Event for each animation that has been created. */
  @EventName("animationCreated")
  EventListener onAnimationCreated(EventHandler<AnimationCreated> eventListener);

  /** Event for animation that has been started. */
  @EventName("animationStarted")
  EventListener onAnimationStarted(EventHandler<AnimationStarted> eventListener);
}

File: pulsar-core/pulsar-tools/pulsar-browser/src/main/kotlin/ai/platon/cdt/kt/protocol/commands/Animation.kt
package ai.platon.cdt.kt.protocol.commands

import ai.platon.cdt.kt.protocol.events.animation.AnimationCanceled
import ai.platon.cdt.kt.protocol.events.animation.AnimationCreated
import ai.platon.cdt.kt.protocol.events.animation.AnimationStarted
import ai.platon.cdt.kt.protocol.support.annotations.EventName
import ai.platon.cdt.kt.protocol.support.annotations.Experimental
import ai.platon.cdt.kt.protocol.support.annotations.ParamName
import ai.platon.cdt.kt.protocol.support.annotations.Returns
import ai.platon.cdt.kt.protocol.support.types.EventHandler
import ai.platon.cdt.kt.protocol.support.types.EventListener
import ai.platon.cdt.kt.protocol.types.runtime.RemoteObject
import kotlin.Boolean
import kotlin.Double
import kotlin.String
import kotlin.Unit
import kotlin.collections.List

@Experimental
public interface Animation {
  /**
   * Disables animation domain notifications.
   */
  public suspend fun disable()

  /**
   * Enables animation domain notifications.
   */
  public suspend fun enable()

  /**
   * Returns the current time of the an animation.
   * @param id Id of animation.
   */
  @Returns("currentTime")
  public suspend fun getCurrentTime(@ParamName("id") id: String): Double

  /**
   * Gets the playback rate of the document timeline.
   */
  @Returns("playbackRate")
  public suspend fun getPlaybackRate(): Double

  /**
   * Releases a set of animations to no longer be manipulated.
   * @param animations List of animation ids to seek.
   */
  public suspend fun releaseAnimations(@ParamName("animations") animations: List<String>)

  /**
   * Gets the remote object of the Animation.
   * @param animationId Animation id.
   */
  @Returns("remoteObject")
  public suspend fun resolveAnimation(@ParamName("animationId") animationId: String): RemoteObject

  /**
   * Seek a set of animations to a particular time within each animation.
   * @param animations List of animation ids to seek.
   * @param currentTime Set the current time of each animation.
   */
  public suspend fun seekAnimations(@ParamName("animations") animations: List<String>,
      @ParamName("currentTime") currentTime: Double)

  /**
   * Sets the paused state of a set of animations.
   * @param animations Animations to set the pause state of.
   * @param paused Paused state to set to.
   */
  public suspend fun setPaused(@ParamName("animations") animations: List<String>,
      @ParamName("paused") paused: Boolean)

  /**
   * Sets the playback rate of the document timeline.
   * @param playbackRate Playback rate for animations on page
   */
  public suspend fun setPlaybackRate(@ParamName("playbackRate") playbackRate: Double)

  /**
   * Sets the timing of an animation node.
   * @param animationId Animation id.
   * @param duration Duration of the animation.
   * @param delay Delay of the animation.
   */
  public suspend fun setTiming(
    @ParamName("animationId") animationId: String,
    @ParamName("duration") duration: Double,
    @ParamName("delay") delay: Double,
  )

  @EventName("animationCanceled")
  public fun onAnimationCanceled(eventListener: EventHandler<AnimationCanceled>): EventListener

  @EventName("animationCanceled")
  public fun onAnimationCanceled(eventListener: suspend (AnimationCanceled) -> Unit): EventListener

  @EventName("animationCreated")
  public fun onAnimationCreated(eventListener: EventHandler<AnimationCreated>): EventListener

  @EventName("animationCreated")
  public fun onAnimationCreated(eventListener: suspend (AnimationCreated) -> Unit): EventListener

  @EventName("animationStarted")
  public fun onAnimationStarted(eventListener: EventHandler<AnimationStarted>): EventListener

  @EventName("animationStarted")
  public fun onAnimationStarted(eventListener: suspend (AnimationStarted) -> Unit): EventListener
}
